lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,57 +1,57 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    attr_reader :batched_queries
-
-    def initialize
-      @batched_queries = Hash.new{ |hash, key| hash[key] = [] }
-    end
-
     class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
       def result
-        result = super
-        until execution_strategy.batched_queries.empty?
-          queries = execution_strategy.batched_queries.shift.last
-          queries.first.class.execute(queries)
-        end
-        result
+        GraphQL::Batch::Promise.resolve(super).sync
+      ensure
+        GraphQL::Batch::Executor.current.clear
       end
     end
 
     class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
       def result
-        result_hash = super
-        result_hash.each do |key, value|
-          if value.is_a?(FieldResolution)
-            value.result_hash = result_hash
+        as_promise_unless_resolved(super)
+      end
+
+      private
+
+      def as_promise_unless_resolved(result)
+        all_promises = []
+        each_promise(result) do |obj, key, promise|
+          obj[key] = nil
+          all_promises << promise.then { |value| obj[key] = value }
+        end
+        return result if all_promises.empty?
+        Promise.all(all_promises).then { result }
+      end
+
+      def each_promise(obj, &block)
+        case obj
+        when Array
+          obj.each_with_index do |value, idx|
+            each_promise_in_entry(obj, idx, value, &block)
+          end
+        when Hash
+          obj.each do |key, value|
+            each_promise_in_entry(obj, key, value, &block)
           end
         end
-        result_hash
+      end
+
+      def each_promise_in_entry(obj, key, value, &block)
+        yield obj, key, value if value.is_a?(::Promise)
+        each_promise(value, &block)
```

Comment:
there's no point in calling this if we yielded above right, should it be in an if/else?